be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_sink_operator.cpp

```diff
@@ -95,41 +103,165 @@ Status SpillableAggregateBlockingSinkOperator::prepare(RuntimeState* state) {
     }
     _peak_revocable_mem_bytes = _unique_metrics->AddHighWaterMarkCounter(
             "PeakRevocableMemoryBytes", TUnit::BYTES, RuntimeProfile::Counter::create_strategy(TUnit::BYTES));
+    _hash_table_spill_times = ADD_COUNTER(_unique_metrics.get(), "HashTableSpillTimes", TUnit::UNIT);
+
+    _ht_low_reduction_threshold = state->agg_spill_ht_low_reduction_threshold();
+    _ht_low_reduction_chunk_limit = state->agg_spill_ht_low_reduction_chunk_limit();
     return Status::OK();
 }
 
 Status SpillableAggregateBlockingSinkOperator::push_chunk(RuntimeState* state, const ChunkPtr& chunk) {
     if (chunk == nullptr || chunk->is_empty()) {
         return Status::OK();
     }
+
+    if (_spill_strategy == spill::SpillStrategy::NO_SPILL) {
+        RETURN_IF_ERROR(AggregateBlockingSinkOperator::push_chunk(state, chunk));
+        set_revocable_mem_bytes(_aggregator->hash_map_memory_usage());
+        return Status::OK();
+    }
+
+    if (state->enable_agg_spill_preaggregation()) {
+        return _try_to_spill_by_auto(state, chunk);
+    } else {
+        return _try_to_spill_by_force(state, chunk);
+    }
+    return Status::OK();
+}
+
+Status SpillableAggregateBlockingSinkOperator::_try_to_spill_by_force(RuntimeState* state, const ChunkPtr& chunk) {
     RETURN_IF_ERROR(AggregateBlockingSinkOperator::push_chunk(state, chunk));
     set_revocable_mem_bytes(_aggregator->hash_map_memory_usage());
-    if (_spill_strategy == spill::SpillStrategy::SPILL_ALL) {
-        return _spill_all_inputs(state, chunk);
+    return _spill_all_data(state, true);
+}
+
+void SpillableAggregateBlockingSinkOperator::_add_non_agg_chunk(ChunkPtr chunk) {
+    _non_agg_rows += chunk->num_rows();
+    _non_agg_bytes += chunk->memory_usage();
+
+    _non_agg_chunks.push(std::move(chunk));
+}
+
+Status SpillableAggregateBlockingSinkOperator::_try_to_spill_by_auto(RuntimeState* state, const ChunkPtr& chunk) {
```

Comment:
can it consider with auto agg? use such window strage do decide pash through, preagg, spill?

Reply:
> can it consider with auto agg? use such window strage do decide pash through, preagg, spill?

The current strategy is to learn from streaming auto agg